builtins/web/form-data/form-data-encoder.cpp

```diff
@@ -0,0 +1,492 @@
+
+  template <typename I> size_t write_and_cache_remainder(StreamContext &stream, I first, I last);
+
+  State next_state(StreamContext &stream);
+  void maybe_drain_leftovers(JSContext *cx, StreamContext &stream);
+  bool handle_entry_header(JSContext *cx, StreamContext &stream);
+  bool handle_entry_body(JSContext *cx, StreamContext &stream);
+  bool handle_entry_footer(JSContext *cx, StreamContext &stream);
+  bool handle_close(JSContext *cx, StreamContext &stream);
+
+public:
+  MultipartFormDataImpl(std::string boundary)
+      : state_(State::Start), boundary_(std::move(boundary)), chunk_idx_(0), file_leftovers_(0) {
+    remainder_.reserve(128);
+  }
+
+  std::string boundary() {  return boundary_; };
+  bool read_next(JSContext *cx, StreamContext &stream);
+};
+
+MultipartFormDataImpl::State MultipartFormDataImpl::next_state(StreamContext &stream) {
+  auto finished = (chunk_idx_ >= stream.entries->length());
+  auto empty = stream.entries->empty();
+
+  switch (state_) {
+  case State::Start:
+    return empty ? State::Done : State::EntryHeader;
+  case State::EntryHeader:
+    return State::EntryBody;
+  case State::EntryBody:
+    return State::EntryFooter;
+  case State::EntryFooter:
+    return finished ? State::Close : State::EntryHeader;
+  case State::Close:
+    return State::Done;
+  case State::Done:
+    return State::Done;
+  default:
+    MOZ_ASSERT_UNREACHABLE("Invalid state");
+  }
+}
+
+void MultipartFormDataImpl::maybe_drain_leftovers(JSContext *cx, StreamContext &stream) {
+  if (!remainder_view_.empty()) {
+    auto written = stream.write(remainder_view_.begin(), remainder_view_.end());
+    remainder_view_.remove_prefix(written);
+
+    if (remainder_view_.empty()) {
+      remainder_.clear();
+      remainder_view_ = remainder_;
+    }
+  }
+
+  if (file_leftovers_ != 0) {
+    auto entry = stream.entries->begin()[chunk_idx_];
+    MOZ_ASSERT(state_ == State::EntryBody);
+    MOZ_ASSERT(File::is_instance(entry.value));
+
+    RootedObject obj(cx, &entry.value.toObject());
+    auto blob = Blob::blob(obj);
+    auto blobsz = blob->length();
+    auto offset = blobsz - file_leftovers_;
+    file_leftovers_ -= stream.write(blob->begin() + offset, blob->end());
+  }
+}
+
+template <typename I>
+size_t MultipartFormDataImpl::write_and_cache_remainder(StreamContext &stream, I first, I last) {
+  auto datasz = static_cast<size_t>(std::distance(first, last));
+  auto written = stream.write(first, last);
+
+  MOZ_ASSERT(written <= datasz);
+
+  auto leftover = datasz - written;
+  if (leftover > 0) {
+    MOZ_ASSERT(remainder_.empty());
+    remainder_.assign(first + written, last);
+    remainder_view_ = remainder_;
+  }
+
+  return written;
+}
+
+bool MultipartFormDataImpl::handle_entry_header(JSContext *cx, StreamContext &stream) {
+  auto entry = stream.entries->begin()[chunk_idx_];
+  auto header = fmt::memory_buffer();
+  auto name = escape_newlines(entry.name).value();
```

Comment:
I think this needs a null check?

Reply:
The function is overloaded but the version that doesn't take `JSContetext` is infallible. I can change this to separate function like `escape_name` and `escape_name_val` if we prefer explicit return value. Anyway, I've added the assertion for now.